test/e2e/mocked_agent.go

```diff
@@ -0,0 +1,198 @@
+func NewMockedAgentSink() *MockedAgentSink {
+	mockedAgentSink := &MockedAgentSink{
+		mtx:               &sync.RWMutex{},
+		eventReceivedChan: make(chan struct{}, 8),
+	}
+	mockedAgentSink.httpServer = httptest.NewServer(mockedAgentSink)
+
+	agentSinkConfig := sinks.SinkConfig{
+		Name: newRelicInfraSinkID,
+		Config: map[string]string{
+			"clusterName":   "integrationTest",
+			"agentEndpoint": "http://" + mockedAgentSink.httpServer.Listener.Addr().String(),
+		},
+	}
+
+	createdSinks, err := sinks.CreateSinks([]sinks.SinkConfig{agentSinkConfig})
+	if err != nil {
+		log.Fatalf("error creating infra sink: %v", err)
+	}
+
+	agentSink, ok := createdSinks[newRelicInfraSinkID]
+	if !ok {
+		log.Fatal("could not retrieve agent infra sink from map")
+	}
+
+	mockedAgentSink.agentSink = agentSink
+
+	return mockedAgentSink
+}
+
+// HandleEvent sends a notification to the event received channel and then forwards it to the underlying sink.
+func (mas *MockedAgentSink) HandleEvent(kubeEvent events.KubeEvent) error {
+	mas.eventReceivedChan <- struct{}{}
+	return mas.agentSink.HandleEvent(kubeEvent)
+}
+
+// ServeHTTP handles a request that would be for the infra-agent and stores the unmarshalled event.
+func (mas *MockedAgentSink) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	mas.mtx.Lock()
+	defer mas.mtx.Unlock()
+
+	var ev struct {
+		Data []struct {
+			Events []sdkEvent.Event `json:"events"`
+		} `json:"data"`
+	}
+
+	defer r.Body.Close() // nolint:errcheck
+	body, _ := ioutil.ReadAll(r.Body)
+	err := json.Unmarshal(body, &ev)
+	if err != nil {
+		log.Fatalf("error getting request body")
+	}
+
+	if len(ev.Data) == 0 {
+		log.Warnf("received payload with no data: %s", string(body))
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	mas.receivedEvents = append(mas.receivedEvents, ev.Data[0].Events...)
+	rw.WriteHeader(http.StatusNoContent) // The the agent does
+}
+
+// Has relaxedly checks whether the mocked agent has received an event.
+func (mas *MockedAgentSink) Has(testEvent *sdkEvent.Event) bool {
+	mas.mtx.RLock()
+	defer mas.mtx.RUnlock()
+
+	for i := range mas.receivedEvents {
+		receivedEvent := &mas.receivedEvents[i]
+
+		if isEventSubset(receivedEvent, testEvent) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Events returns the list of events the mock has captured.
+func (mas *MockedAgentSink) Events() []sdkEvent.Event {
+	mas.mtx.RLock()
+	defer mas.mtx.RUnlock()
+
+	retEvents := make([]sdkEvent.Event, len(mas.receivedEvents))
+	copy(retEvents, mas.receivedEvents)
+
+	return retEvents
+}
+
+// ForgetEvents erases all the recorded events.
+func (mas *MockedAgentSink) ForgetEvents() {
+	mas.mtx.Lock()
+	defer mas.mtx.Unlock()
+
+	mas.receivedEvents = nil
+}
+
+// Wait blocks until betweenEvents time has passed since the last received event, or up to max time has passed since the call.
+// Returns false if we had to exhaust max.
+func (mas *MockedAgentSink) Wait(betweenEvents, max time.Duration) bool {
```

Comment:
are we checking the return value?

Reply:
Nop! I just thought it woudl be useful information (similar to context and time.Timer), but so far it's not used since the test is going to fail anyway if we didn't catch the event.